internal/core/software_renderer.rs

```diff
@@ -314,12 +314,76 @@ impl Renderer for SoftwareRenderer {
+        byte_offset: usize,
     ) -> LogicalRect {
-        Default::default()
+        let window_adapter = match self.window.upgrade() {
+            Some(window) => window,
+            None => return Default::default(),
+        };
+
+        let scale_factor = ScaleFactor::new(window_adapter.window().scale_factor());
+        let string = text_input.text();
+
+        let font_request = text_input.font_request(&window_adapter);
+        let font = fonts::match_font(&font_request, scale_factor);
+
+        let width: Length<i16, crate::lengths::PhysicalPx> =
+            Length::new((text_input.width().0 * scale_factor.0).round() as i16);
+        let height: Length<i16, crate::lengths::PhysicalPx> =
+            Length::new((text_input.height().0 * scale_factor.0).round() as i16);
+
+        let (cursor_position, cursor_height) = match font {
+            fonts::Font::PixelFont(pf) => {
+                let layout = fonts::text_layout_for_font(&pf, &font_request, scale_factor);
+
+                let paragraph = TextParagraphLayout {
+                    string: &string,
+                    layout,
+                    max_width: width,
+                    max_height: height,
+                    horizontal_alignment: text_input.horizontal_alignment(),
+                    vertical_alignment: text_input.vertical_alignment(),
+                    wrap: text_input.wrap(),
+                    overflow: TextOverflow::Clip,
+                    single_line: false,
+                };
+
+                (paragraph.cursor_pos_for_byte_offset(byte_offset), pf.height())
+            }
+            #[cfg(feature = "software-renderer-systemfonts")]
+            fonts::Font::VectorFont(vf) => {
+                let layout = fonts::text_layout_for_font(&vf, &font_request, scale_factor);
+
+                let paragraph = TextParagraphLayout {
+                    string: &string,
+                    layout,
+                    max_width: width,
+                    max_height: height,
+                    horizontal_alignment: text_input.horizontal_alignment(),
+                    vertical_alignment: text_input.vertical_alignment(),
+                    wrap: text_input.wrap(),
+                    overflow: TextOverflow::Clip,
+                    single_line: false,
+                };
+
+                (paragraph.cursor_pos_for_byte_offset(byte_offset), vf.height())
+            }
+        };
+
+        let cursor_x = cursor_position.0.cast::<f32>().cast_unit::<crate::lengths::LogicalPx>().0;
+        let cursor_y = cursor_position.1.cast::<f32>().cast_unit::<crate::lengths::LogicalPx>().0;
+
+        let cursor_width = text_input.text_cursor_width().0 * scale_factor.0;
+        let cursor_height = cursor_height.cast::<f32>().cast_unit::<crate::lengths::LogicalPx>().0;
+
+        return LogicalRect::new(
+            LogicalPoint::new(cursor_x, cursor_y),
+            LogicalSize::new(cursor_width, cursor_height),
+        );
```

Comment:
Prefer `XXX::from_lengths()` over `XXX::new()` where possible, as the former will tell if you the types are wrong, while the latter will not.

```suggestion
        return PhysicalRect::new(
            PhysicalPoint::from_lengths(cursor_position.0, cursor_position.0),
            PhysicalSize::from_lengths(
                (text_input.text_cursor_width() * scale_factor).cast(),
                cursor_height,
            ),
        )
        .cast()
            / scale_factor;
```